docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,309 @@
+  // Watch registers the node for updates to descriptor versions. A
+  // watch should be registered before any other lease operation.
+  rpc Watch(WatchRequest) returns (stream WatchResponse) {}
+  
+  // Acquire acquires a read lease for the specified descriptor and
+  // version. The lease is associated with the specified node. An
+  // error will be returned if the specified version is not the most
+  // recent version for the descriptor.
+  rpc Acquire(AcquireRequest) returns (AcquireResponse) {}
+
+  // Release allows a read lease to be released before it
+  // expires. This should be used when a watch request indicates a new
+  // version of a descriptor is available and the lease holder
+  // determines it is no longer using the old version.
+  rpc Release(ReleaseRequest) returns (ReleaseResponse) {}
+
+  // Publish publishes a table descriptor change, writing the table
+  // descriptor to the KV store on success and notifying all watches
+  // of the new version. An error is returned if the version of the
+  // table descriptor is not equal to the current version + 1. An
+  // error is returned if there are two active versions of the table
+  // descriptor.
+  rpc Publish(PublishRequest) returns (PublishResponse) {}
+}
+```
+
+When a node starts up it will register a `Watch` with the lease master
+in order to receive a stream of version updates for all
+descriptors. Each node will maintain a map from `<descriptorID,
+version>` to `<TableDescriptor, expiration, localRefCount>`. When the
+node needs to use a descriptor for an operation it will look in its
+local map. If the descriptor is not there it will be read from the KV
+store and a lease will be acquired by sending an `Acquire` message to
+the lease master. The node will maintain the invariant that a new
+reference to a descriptor will only be added to the most recent
+version. When the reference count for an older version falls to 0 the
+node will send a `Release` message to the lease master.
+
+The `Watch` responses will notify the node when a new version of a
+descriptor is available which it will then read from the KV store if
+the node has a lease for the descriptor.
+
+When performing a schema change operation, the modification to the
+table descriptor is performed using the `Publish` operation which will
+fail if the version is invalid (i.e. too old) or if there are already
+two active versions of the descriptor in the cluster. On success, the
+`Publish` operation will write the new descriptor to the KV store and
+notify all watches of the new descriptor which will cause those nodes
+to stop using the old descriptor for new operations. As existing
+operations complete they will release their local reference to the old
+descriptor which will cause the reference count to drop to 0 and
+result in the lease being released. When all of the leases for the old
+descriptor have been released, the lease master will send out a
+notification to the watches that the version is `DEAD`.
```

Comment:
The intention was that schema change operation would watch for `DEAD` notifications so that it could proceed with the next step. In the description about adding an index I talk about waiting for a change to propagate to all nodes. In reality, the schema change operation is waiting for the previous version to no longer be in use by any nodes. I should make that clearer.